### ISSUE TITLE
Patching or  installing GmsCore in priv-app

### ISSUE THREAD

**Issue by pelomedusa** [microg/android_packages_apps_GmsCore#462]
> Note: On Android 7 (or later) an additional patch is needed to make location work, or alternatively, you can install GmsCore.apk in the /system/priv-app folder.

I'm running LineageOS 15. I managed to follow your installation guide from the beginning to the end, except for the very last part (quoted above).
Are there any detailed instruction for this anywhere? I'm not really sure how to apply this patch or how to move GmsCore to the right location.
Thanks for your work, it's amazing.

**Comment by ArchangeGabriel** [comment]
Indeed, the whole documentation could need a revamp, but basically:
1. Remount system partition as `rw`, see https://android.stackexchange.com/questions/110927/how-to-mount-system-rewritable-or-read-only-rw-ro for instance.
2. Run `adb push GmsCore.apk /system/priv-app`.
3. Remount system as `ro`.
4. Reboot.

They are other ways (I know some apps allow doing it from the system —with root right ofc). And if you can wait for official LineageOS 15 builds, then you would just need to install the one from lineage.microg.org.

**Comment by Shadow53** [comment]
> Run `adb push GmsCore.apk /system/priv-app`.

To make everything work correctly, you'll want to push the apk to `/system/priv-app/GmsCore/GmsCore.apk` and then extract the `lib/` folder from the apk to `/system/priv-app/GmsCore/lib` - otherwise the bundled libraries will not be found by the app and the maps library will crash apps.

**Comment by Roboe** [comment]
@ArchangeGabriel Note that [priv-apps need a special XML permissions manifest](https://source.android.com/devices/tech/config/perms-whitelist) from Android Oreo onwards. You will need to push the appropiate for [GmsCore](https://github.com/lineageos4microg/android_prebuilts_prebuiltapks/blob/master/GmsCore/privapp-permissions-com.google.android.gms.xml) and [FakeStore](https://github.com/lineageos4microg/android_prebuilts_prebuiltapks/blob/af3476814391295e07f9e497470fc01002993c0f/FakeStore/privapp-permissions-com.android.vending.xml) to `/etc/permissions` too.

> otherwise the bundled libraries will not be found by the app and the maps library will crash apps.

This is not true since Maps v2 API doesn't need bundled libraries. You'll need to push Maps v1 API `jar` library if you want to support those, though.

**Comment by ale5000-git** [comment]
@Roboe: You misunderstood Shadow53, Maps v2 API need the library `libvtm-jni.so` (it is inside GmsCore.apk, you can see by opening it with a zip decompression utility).
When GmsCore is installed as normal app the library is installed automatically from Android but if you install it as system app you need to do it manually.

But there is a note about this, from about Marshmallow and up Android can use automatically libraries inside apk files without decompressing them (since GmsCore has the proper flag) but this feature isn't granted to be present in all ROMs so it is better to extract the library just to be safe.

These things apply to all system apps that contain binary libraries.

**Comment by Roboe** [comment]
Thanks for the clarification. I've been flashing my own ZIP of microG apps since Lollipop and have not had any issues.

So, for the OP issue: it's not needed for Android Oreo, but should be done on pre-Marshmallow releases.

**Comment by ale5000-git** [comment]
This feature if I'm not wrong was introduced by Google but wasn't in AOSP.
Then CyanogenMod team got it implemented but I'm unsure if all Marshmallow ROMs have it.

Also for this feature to be working the app must have in the manifest `android:extractNativeLibs="false"` and must be zip-aligned (GmsCore comply with this but other apps maybe be not).
